src/xcode/ENA/ENA/Source/Services/HealthCertificateValidationService/HealthCertificateValidationService.swift

```diff
@@ -172,4 +212,557 @@ final class HealthCertificateValidationService: HealthCertificateValidationProvi
+			return
+		}
+		
+		Log.info("Successfully verified eTag. Proceed with package extraction...", log: .vaccination)
+				
+		guard !packageDownloadResponse.isEmpty,
+			  let sapDownloadedPackage = packageDownloadResponse.package else {
+			Log.error("PackageDownloadResponse is empty. Return with failure.", log: .vaccination)
+			completion(.failure(.ACCEPTANCE_RULE_JSON_ARCHIVE_FILE_MISSING))
+			return
+		}
+		Log.info("Successfully extracted sapDownloadedPackage. Proceed with package verification...", log: .vaccination)
+		
+		guard self.signatureVerifier.verify(sapDownloadedPackage) else {
+			Log.error("Verification of sapDownloadedPackage failed. Return with failure", log: .vaccination)
+			completion(.failure(.ACCEPTANCE_RULE_JSON_ARCHIVE_SIGNATURE_INVALID))
+			return
+		}
+		Log.info("Successfully verified sapDownloadedPackage. Proceed now with CBOR decoding...", log: .vaccination)
+		
+		
+		self.acceptanceRules(sapDownloadedPackage.bin, completion: { result in
+			switch result {
+			case let .success(acceptanceRules):
+				Log.info("Successfully decoded acceptance rules: \(private: acceptanceRules).", log: .vaccination)
+				// Save in success case for caching
+				let receivedAcceptanceRules = ValidationRulesCache(
+					validationRules: acceptanceRules,
+					lastValidationRulesETag: eTag
+				)
+				store.acceptanceRulesCache = receivedAcceptanceRules
+				Log.info("Successfully stored acceptance rules in cache.", log: .vaccination)
+				completion(.success(acceptanceRules))
+			case let .failure(error):
+				Log.error("Could not decode CBOR from package with error:", log: .vaccination, error: error)
+				completion(.failure(error))
+			}
+		})
+	}
+	
+	private func acceptanceRuleDownloadingFailureHandler(
+		error: URLSession.Response.Failure,
+		completion: @escaping (Result<[Rule], HealthCertificateValidationError>) -> Void
+	) {
+		switch error {
+		case .notModified:
+			// Normally we should have cached something before
+			Log.info("Download new acceptance rules aborted due to not modified content. Taking cached rules.", log: .vaccination)
+			if let cachedAcceptanceRules = store.acceptanceRulesCache?.validationRules {
+				completion(.success(cachedAcceptanceRules))
+			} else {
+				// If not, return edge case error
+				Log.error("Could not find cached acceptance rules but need some.", log: .vaccination)
+				completion(.failure(.ACCEPTANCE_RULE_MISSING_CACHE))
+			}
+		case .noNetworkConnection:
+			Log.error("Could not download acceptance rules due to no network.", log: .vaccination, error: error)
+			completion(.failure(.NO_NETWORK))
+		case let .serverError(statusCode):
+			switch statusCode {
+			case 400...409:
+				Log.error("Could not download acceptance rules due to client error.", log: .vaccination, error: error)
+				completion(.failure(.ACCEPTANCE_RULE_CLIENT_ERROR))
+			default:
+				Log.error("Could not download acceptance rules due to server error.", log: .vaccination, error: error)
```

Comment:
Maybe we could log the status code here as well

Reply:
Good idea here, too ;)